BMS_Software_Main/BMS_Software_Main.ino

```diff
@@ -0,0 +1,170 @@
+}
+
+void setInputPins()
+{
+  pinMode(PACK_I_MEAS_PIN,      INPUT);
+  pinMode(PACK_V_MEAS_PIN,      INPUT);
+  pinMode(LOGIC_V_MEAS_PIN,     INPUT);
+  pinMode(TEMP_degC_MEAS_PIN,   INPUT);
+  pinMode(C1_V_MEAS_PIN,        INPUT);
+  pinMode(C2_V_MEAS_PIN,        INPUT);
+  pinMode(C3_V_MEAS_PIN,        INPUT);
+  pinMode(C4_V_MEAS_PIN,        INPUT);
+  pinMode(C5_V_MEAS_PIN,        INPUT);
+  pinMode(C6_V_MEAS_PIN,        INPUT);
+  pinMode(C7_V_MEAS_PIN,        INPUT);
+  pinMode(C8_V_MEAS_PIN,        INPUT);
+}
+
+void setOutputPins()
+{
+  pinMode(PACK_OUT_CTRL_PIN,     OUTPUT);
+  pinMode(LOGIC_SWITCH_CTRL_PIN, OUTPUT);
+  pinMode(BUZZER_CTRL_PIN,       OUTPUT);
+  pinMode(FAN_CTRL_PIN,          OUTPUT);
+  pinMode(FAN_PWR_IND_PIN,      OUTPUT);
+  pinMode(SW_IND_PIN,           OUTPUT);
+  pinMode(SW_ERR_IND_PIN,       OUTPUT);
+}
+
+void setOutputStates() //is the pin on or off? must be set for all digital pins
+{
+  digitalWrite(PACK_OUT_CTRL_PIN,      HIGH);
+  digitalWrite(LOGIC_SWITCH_CTRL_PIN,  LOW);
+  digitalWrite(BUZZER_CTRL_PIN,        LOW);
+  digitalWrite(FAN_CTRL_PIN,           LOW);
+  digitalWrite(FAN_PWR_IND_PIN,       LOW);
+  digitalWrite(SW_IND_PIN,            LOW);
+  digitalWrite(SW_ERR_IND_PIN,        LOW);
+}
+
+
+void estop(uint8_t time)
+{
+  if (time > 0)
+  {
+    digitalWrite(PACK_OUT_CTRL_PIN, LOW);
+    delay(time);
+    digitalWrite(PACK_OUT_CTRL_PIN, HIGH);
+  }
+}
+
+// get temp for battery pack and if too high, turn on fans
+float getTemperature()
+{
+  temp = analog.read(TEMP_degC_MEAS_PIN);
+  if (temp > TEMP_THRESHOLD)
+  {
+    delay(DEBOUNCE_DELAY);
+    temp = analog.read(TEMP_degC_MEAS_PIN);
+    if (temp >= TEMP_THRESHOLD)
+    {
+      uint8_t superhot = 1;
+      RoveComm.write(RC_BMSBOARD_PACKSUPERHOT_DATA_ID, superhot);
+    }
+  }
+  return temp/1000;
+}
+
+// get the individual voltage of 8 cells and put in an array
+float* getCellVoltages()
+{
+  float cells[RC_BMSBOARD_CELLV_MEAS_DATA_COUNT] = {};
+
+  for a in range(RC_BMSBOARD_CELLV_MEAS_DATA_COUNT)
+  {
+    float cell = map(analogRead(CELL_MEAS_PINS[a]), PACK_V_ADC_MIN, PACK_V_ADC_MAX, VOLTS_MIN, PACK_VOLTS_MAX);
+    // check errors
+    if(cell =< CELL_UNDERVOLTAGE)
+    {
+      delay(DEBOUNCE_DELAY);
+      float cell = map(analogRead(CELL_MEAS_PINS[a]), PACK_V_ADC_MIN, PACK_V_ADC_MAX, VOLTS_MIN, PACK_VOLTS_MAX);
+      if (cell =< CELL_UNDERVOLTAGE)
+      {
+        uint8_t undervoltcell = a;  
+        RoveComm.write(RC_BMSBOARD_CELLUNDERVOLTAGE_DATA_ID, undervoltcell);
+      }
+    }
+    cells[a] = cell/1000;
+  }
+  return float cells[RC_BMSBOARD_CELLV_MEAS_DATA_COUNT];
+}
+
+float getPackVoltage(float cells[RC_BMSBOARD_CELLV_MEAS_DATA_COUNT])
+{
+  packVoltage = 0;
+  for a in range(RC_BMSBOARD_CELLV_MEAS_DATA_COUNT)
+  {
+    packVoltage += a;
+  }
+  if (packVoltage =< PACK_UNDERVOLTAGE)
+  {
+    uint8_t packerror = 1;
+    RoveComm.write(RC_BMSBOARD_PACKUNDERVOLTAGE_DATA_ID,packerror);
+  }
+  return packVoltage/1000;
```

Comment:
Do some checking and leave some comments to make sure if the /1000 is necessary or not. Kinda hard to follow if this is V or mV